proposals/stack-switching/Explainer.md

```diff
@@ -0,0 +1,876 @@
+
+- `cont <typeidx>` is a new form of composite type
+  - `(cont $ft) ok` iff `$ft ok` and `$ft = [t1*] -> [t2*]`
+
+We add two new continuation heap types and their subtyping hierarchy:
+- `heaptypes ::= ... | cont | nocont`
+- `nocont ok` and `cont ok` always
+- `nocont` is the bottom type of continuation types, whereas `cont` is the top type, i.e. `nocont <: cont`
+
+### Tags
+
+We change the wellformedness condition for tag types to be more liberal, i.e.
+
+- `(tag $t (type $ft)) ok` iff `$ft ok` and `$ft = [t1*] -> [t2*]`
+
+In other words, the return type of tag types is allowed to be non-empty.
+
+### Instructions
+
+The new instructions and their validation rules are as follows. To simplify the presentation, we write this:
+
+```
+C.types[$ct] ~~ cont [t1*] -> [t2*]
+```
+
+where we really mean this:
+
+```
+C.types[$ct] ~~ cont $ft
+C.types[$ft] ~~ func [t1*] -> [t2*]
+```
+
+This abbreviation will be formalised with an auxiliary function or other means in the spec.
+
+- `cont.new <typeidx>`
+  - Create a new continuation from a given typed funcref.
+  - `cont.new $ct : [(ref null? $ft)] -> [(ref $ct)]`
+    - iff `C.types[$ct] ~~ cont [t1*] -> [t2*]`
+
+- `cont.bind <typeidx> <typeidx>`
+  - Partially apply a continuation.
+  - `cont.bind $ct $ct' : [t3* (ref null? $ct)] -> [(ref $ct')]`
+    - iff `C.types[$ct] ~~ cont [t3* t1*] -> [t2*]`
+    - and `C.types[$ct'] ~~ cont [t1'*] -> [t2'*]`
+    - and `[t1*] -> [t2*] <: [t1'*] -> [t2'*]`
+
+- `resume <typeidx> hdl*`
+  - Execute a given continuation.
+    - If the executed continuation suspends with a control tag `$t`, the corresponding handler `(on $t H)` is executed.
+  - `resume $ct hdl* : [t1* (ref null? $ct)] -> [t2*]`
+    - iff `C.types[$ct] ~~ cont [t1*] -> [t2*]`
+    - and `(hdl : t2*)*`
+
+- `resume_throw <typeidx> <exnidx> hdl*`
+  - Execute a given continuation, but force it to immediately throw the annotated exception.
+  - Used to abort a continuation.
+  - `resume_throw $ct $e hdl* : [te* (ref null? $ct)] -> [t2*]`
+    - iff `C.types[$ct] ~~ cont [t1*] -> [t2*]`
+    - and `C.tags[$e] : tag $ft`
+    - and `C.types[$ft] ~~ func [te*] -> []`
+    - and `(hdl : t2*)*`
+
+- `hdl = (on <tagidx> <labelidx>) | (on <tagidx> switch)`
+  - Handlers attached to `resume` and `resume_throw`, handling control tags for `suspend` and `switch`, respectively.
+  - `(on $e $l) : t*`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
+    - and `C.labels[$l] = [t1'* (ref null? $ct)]`
+    - and `t1* <: t1'*`
+    - and `C.types[$ct] ~~ cont [t2'*] -> [t'*]`
+    - and `[t2*] -> [t*] <: [t2'*] -> [t'*]`
+  - `(on $e switch) : t*`
+    - iff `C.tags[$e] = tag $ft`
+    - and `C.types[$ft] ~~ func [] -> [t*]`
+
+- `suspend <tagidx>`
+  - Use a control tag to suspend the current computation.
+  - `suspend $t : [t1*] -> [t2*]`
+    - iff `C.tags[$t] = tag $ft`
+    - and `C.types[$ft] ~~ func [t1*] -> [t2*]`
+
+- `switch <typeidx> <typeidx> <tagidx>`
```

Comment:
I know this was mentioned at the meeting this morning, but I missed the reason for having both types annotated. Can we go back to having just `$ct1`?

Reply:
This is something we can definitely change later, but unless you think this change is vital now, my current inclination is to leave it as is for tomorrow's vote. (My understanding is that the design isn't cut and dried here and e.g. @rossberg may favour the current design.)

Reply:
I'd prefer to fix it sooner rather than later to minimize unnecessary churn once more projects have started implementing the proposal, but I agree it's not critical. I can't imagine a reason for wanting to keep the second annotation, though. It's not necessary for principle typing or any other established type system properties AFAICS.

Reply:
Agreed that we should resolve this sooner rather than later. Let's prioritise converging on this as soon as possible after today's CG meeting. I think this should be pretty easy to resolve, but let's discuss it in a separate issue.